spring-batch-integration/src/main/java/org/springframework/batch/integration/partition/RemotePartitioningMasterStepBuilder.java

```diff
@@ -0,0 +1,297 @@
+		Assert.notNull(jobExplorer, "jobExplorer must not be null");
+		this.jobExplorer = jobExplorer;
+		return this;
+	}
+
+	/**
+	 * How often to poll the job repository for the status of the workers. Defaults to 10 seconds.
+	 * @param pollInterval the poll interval value in milliseconds
+	 * @return this builder instance for fluent chaining
+	 */
+	public RemotePartitioningMasterStepBuilder pollInterval(long pollInterval) {
+		Assert.isTrue(pollInterval > 0, "The poll interval must be greater than zero");
+		this.pollInterval = pollInterval;
+		return this;
+	}
+
+	/**
+	 * When using job repository polling, the time limit to wait. Defaults to -1 (no timeout).
+	 * @param timeout the timeout value in milliseconds
+	 * @return this builder instance for fluent chaining
+	 */
+	public RemotePartitioningMasterStepBuilder timeout(long timeout) {
+		this.timeout = timeout;
+		return this;
+	}
+
+	/**
+	 * Set the bean factory.
+	 * @param beanFactory the bean factory to use
+	 * @return this builder instance for fluent chaining
+	 */
+	public RemotePartitioningMasterStepBuilder beanFactory(BeanFactory beanFactory) {
+		this.beanFactory = beanFactory;
+		return this;
+	}
+
+	public Step build() {
+		Assert.notNull(this.outputChannel, "An outputChannel must be provided");
+
+		// configure messaging template
+		if (this.messagingTemplate == null) {
+			this.messagingTemplate = new MessagingTemplate();
+		}
+		this.messagingTemplate.setDefaultChannel(this.outputChannel);
+
+		// Configure the partition handler
+		final MessageChannelPartitionHandler partitionHandler = new MessageChannelPartitionHandler();
+		partitionHandler.setStepName(getStepName());
+		partitionHandler.setGridSize(getGridSize());
+		partitionHandler.setMessagingOperations(this.messagingTemplate);
+
+		if (isPolling()) {
+			partitionHandler.setJobExplorer(this.jobExplorer);
+			partitionHandler.setPollInterval(this.pollInterval);
+			partitionHandler.setTimeout(this.timeout);
+		}
+		else {
+			PollableChannel replies = new QueueChannel();
+			partitionHandler.setReplyChannel(replies);
+			StandardIntegrationFlow standardIntegrationFlow = IntegrationFlows
+					.from(this.inputChannel)
+					.aggregate(aggregatorSpec -> aggregatorSpec.processor(partitionHandler))
+					.channel(replies)
+					.get();
+			IntegrationFlowContext integrationFlowContext = this.beanFactory.getBean(IntegrationFlowContext.class);
+			integrationFlowContext.registration(standardIntegrationFlow)
+					.autoStartup(false)
+					.register();
+		}
+
+		try {
+			partitionHandler.afterPropertiesSet();
+			super.partitionHandler(partitionHandler);
+		}
+		catch (Exception e) {
+			throw new BeanCreationException("Unable to create a master step for remote partitioning", e);
+		}
+
+		return super.build();
+	}
+
+	private boolean isPolling() {
+		return this.inputChannel == null;
+	}
+
+	@Override
+	public RemotePartitioningMasterStepBuilder repository(JobRepository jobRepository) {
```

Comment:
Yeah... In Spring Integration Java DSL to have an inheritance for builders and to avoid such a duplication we use a concretization via generic argument:
```
public class AggregatorSpec extends CorrelationHandlerSpec<AggregatorSpec, AggregatingMessageHandler> {
``` 
for example.